### DISCUSSION TITLE
Expose named parameters support in the C API

### DISCUSSION THREAD

**Discussion by Giorgi** [duckdb/duckdb#7076] [Ideas]
Currently, named parameters are supported by the Python library only: [[Python] Add support for named parameters #5611 ](https://github.com/duckdb/duckdb/pull/5611)

If you expose an API to support named parameters in the C API it will be possible to implement them in other client libraries, for example in the .Net library: [DbParameter Support](https://github.com/Giorgi/DuckDB.NET/issues/69)

**Comment by Tishj** [comment]
The support is added to the parser, the only thing missing is mapping the named parameters to the right positional parameters.
(taken from the python api code)
```c++
	auto &named_param_map = prep->named_param_map;
	if (py::isinstance<py::dict>(params)) {
		if (named_param_map.empty()) {
			throw InvalidInputException("Param is of type 'dict', but no named parameters were found in the query");
		}
		// Transform named parameters to regular positional parameters
		params = TransformNamedParameters(named_param_map, params);
		// Clear the map, we don't need it anymore
		prep->named_param_map.clear();
	} else if (!named_param_map.empty()) {
		throw InvalidInputException("Named parameters found, but param is not of type 'dict'");
	}
```
TransformNamedParameters just maps the incoming parameters (`dict[str, object]` in Python) to the positional parameters, using the `PreparedStatement::named_param_map`

**Comment by Giorgi** [comment]
I think the simplest option is to add an API that returns the index for a given parameter. This way all the `duckdb_bind_XXX` can stay as they are. Similar to [sqlite3_bind_parameter_index](https://www.sqlite.org/c3ref/bind_parameter_index.html)

**Reply by Tishj** [reply]
That function is simple enough to add, the main difficulty here is that not all clients can support it, and I'd like to still be able to explicitly throw when named parameters are used when they are not supported.
I guess we could add a pragma/setting to disable this check, essentially making the promise that you check that the named parameters are provided properly.

Maybe it should actually be a parameter to the DBConfig constructor, as not to expose this setting to users - because it doesn't make any sense for them to be able to change this.

**Reply by Tishj** [reply]
I was working on adding this option, but I really didnt like how clients should set this, and we would essentially be creating an option that only clients should interact with - while still grouping this under user-interactible options (similar to `allow_external_access` I guess)

So i took another look at one of the clients that I thought was the main blocker for implementing this in core: the CLI - and I came up with a solution that should allow us to support named parameters in core.

**Comment by Giorgi** [comment]
This is now implemented, thanks!
